sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledSymmetricHashJoinExec.scala

```diff
@@ -1063,14 +1121,46 @@ class BigInnerJoinIterator(
     subIter.get.next()
   }
 
+  override def close(): Unit = {
+    super.close()
+    buildSideRowTrackers.flatten.foreach(_.close())
+    isExhausted = true
+  }
+
   private def setupNextJoinIterator(): Unit = {
     while (!isExhausted && !subIter.exists(_.hasNext)) {
-      if (nextJoinGroupIndex >= joinGroups.length) {
+      if (info.joinType == FullOuter) {
+        // save off the build side tracker buffer for the join group just processed
+        subIter match {
+          case Some(streamIter: HashFullJoinStreamSideIterator) =>
+            assert(buildSideRowTrackers(currentJoinGroupIndex).isEmpty, "unexpected row tracker")
+            buildSideRowTrackers(currentJoinGroupIndex) = streamIter.releaseBuiltSideTracker()
+            streamIter.close()
+          case _ =>
+        }
+      }
+      if (currentJoinGroupIndex >= joinGroups.length - 1) {
         // try to pull in the next stream batch
         if (streamPartitioner.hasInputBatches) {
           streamPartitioner.partitionNextBatch()
-          nextJoinGroupIndex = 0
           subIter = Some(moveToNextBuildGroup())
+        } else if (info.joinType == FullOuter) {
+          currentJoinGroupIndex = buildSideRowTrackers.indexWhere(_.isDefined)
+          if (currentJoinGroupIndex == -1) {
+            isExhausted = true
+            subIter = None
+          } else {
+            // Setup an iterator to produce the final full outer join batches for the join group.
+            // TODO: Can free the build-side batch in the build partitioner early here since
+            //       this will be the last iterator to use it.
+            val tracker = buildSideRowTrackers(currentJoinGroupIndex)
+            buildSideRowTrackers(currentJoinGroupIndex) = None
+            subIter = Some(new HashFullJoinIterator(
+              buildPartitioner.getBuildBatch(currentJoinGroupIndex), info.exprs.boundBuildKeys,
+              tracker, Iterator.empty, info.exprs.boundStreamKeys, info.exprs.streamOutput,
+              None, 0, gpuBatchSizeBytes, info.buildSide, info.exprs.compareNullsEqual,
```

Comment:
We don't need the inequality condition because the stream iterator is empty.  There's no join condition to evaluate as there's no row pairs to evaluate the condition.  We're just reusing the existing HashFullJoinIterator code only to produce the final batch consisting of rows in the build side that didn't match any rows in the stream side.  I'll update the comment here to make this more clear.

Reply:
ah ok I missed this, thank you for the comment